### PR TITLE
cd9660: Add support for mask,dirmask,uid,gid options

### PR DESCRIPTION
**Pull request by ricardobranco777**
This patch adds arbitrary mask, dirmask, uid & gid support to ISO9660 ala MSDOSFS.

To test:

```
cd /usr/src/sbin/mount_cd9660
make clean obj depend
make
sudo make install
cd /usr/src/sys/modules/cd9660
make clean obj depend
make
sudo make install
sudo kldunload cd9660
sudo mdconfig -a -t vnode -f /path/to/iso.iso
sudo mount_cd9660 -m640 -M750 -Uricardo -Goperator /dev/md0 /mnt
```

### CODE REVIEW

File: sbin/mount_cd9660/mount_cd9660.8 [hunk 1]
```diff
@@ -66,6 +70,37 @@ Do not strip version numbers on files.
...
+.It Fl m Ar mask
+Specify the maximum file permissions for files
+in the file system.
+(For example, a
+.Ar mask
+of
+.Li 755
+specifies that, by default, the owner should have
+read, write, and execute permissions for files, but
+others should only have read and execute permissions.
```
Comment (jlduran) [inline sbin/mount_cd9660/mount_cd9660.8:90]: Missing closing parenthesis.

File: sbin/mount_cd9660/mount_cd9660.8
Comment (jlduran) [inline sbin/mount_cd9660/mount_cd9660.8:1]: Missing date bump.

**Review by jlduran — COMMENTED** [review]
Nice! Just a couple of minor issues with the man page.

File: sbin/mount_cd9660/mount_cd9660.8 (hunk 1 above)
Comment (ricardobranco777) [reply sbin/mount_cd9660/mount_cd9660.8:90]: > Missing closing parenthesis.

Good catch.  I copied this verbatim from the `mount_msdosfs` manpage that has the same error.

File: sbin/mount_cd9660/mount_cd9660.8
Comment (ricardobranco777) [reply sbin/mount_cd9660/mount_cd9660.8:1]: Fixed.

File: sbin/mount_cd9660/mount_cd9660.c [hunk 2]
```diff
@@ -254,3 +272,58 @@ set_charset(struct iovec **iov, int *iovlen, const char *localcs)
...
+	return (uid);
+}
+
+static mode_t
+a_mask(const char *s)
+{
+	int done, rv;
+	char *ep;
+
+	done = 0;
```
Comment (bsdjhb) [inline sbin/mount_cd9660/mount_cd9660.c:320]: You might be able to simplify this slightly as:

```
     rv = strtol(optarg, &ep, 8);
     if (ep == optarg || rv < 0 || *ep != '\0')
         errx(...)
```

That is, relying on strtol to return optarg == ep if the first character is invalid and not needing `done` or a default value for `rv`.

Note that FreeBSD's style prefers explicit tests for things like `foo != NULL` or `bar != '\0'` rather than relying on implicit tests like `if (foo)`.

File: sys/fs/cd9660/cd9660_node.c [hunk 3]
```diff
@@ -146,8 +146,22 @@ cd9660_defattr(struct iso_directory_record *isodir, struct iso_node *inop,
 	}
 	if (!ap) {
 		inop->inode.iso_mode |= S_IRUSR|S_IXUSR|S_IRGRP|S_IXGRP|S_IROTH|S_IXOTH;
-		inop->inode.iso_uid = (uid_t)0;
-		inop->inode.iso_gid = (gid_t)0;
+		/*
```
Comment (bsdjhb) [inline sys/fs/cd9660/cd9660_node.c:149]: The changes in this function should be a separate commit with its own log message.

File: sys/fs/cd9660/cd9660_node.c [hunk 4]
```diff
@@ -146,8 +146,22 @@ cd9660_defattr(struct iso_directory_record *isodir, struct iso_node *inop,
...
+		 * Set some "sane" permissions for non-Rockridge volumes
+		 */
+		switch (inop->inode.iso_mode & S_IFMT) {
+		case S_IFREG:
+			/* Turn off executable bits for regular files */
+			inop->inode.iso_mode &= ~(VEXEC | VEXEC >> 3 | VEXEC >> 6);
```
Comment (bsdjhb) [inline sys/fs/cd9660/cd9660_node.c:155]: I would suggest using `S_IX*` instead of `VEXEC` with shifts.  However, I'm not sure if makes sense to turn off executable permissions on all files as a non-rock-ridge image may contain executables.

File: sys/fs/cd9660/cd9660_node.c [hunk 5]
```diff
@@ -146,8 +146,22 @@ cd9660_defattr(struct iso_directory_record *isodir, struct iso_node *inop,
...
+			/* FALLTHROUGH */
+		case S_IFDIR:
+			/* Turn on write permissions (owner) */
+			inop->inode.iso_mode |= VWRITE;
```
Comment (bsdjhb) [inline sys/fs/cd9660/cd9660_node.c:159]: Should be `S_IWUSR` instead of `VWRITE`, but also cd9660 in FreeBSD doesn't support read/write mounts so I'm not sure why it makes sense to add this permission for regular files and directories?

File: sys/fs/cd9660/cd9660_vfsops.c [hunk 6]
```diff
@@ -394,6 +403,23 @@ iso_mountfs(struct vnode *devvp, struct mount *mp)
 	vfs_flagopt(mp->mnt_optnew, "nojoliet", &isomp->im_flags, ISOFSMNT_NOJOLIET);
 	vfs_flagopt(mp->mnt_optnew, "kiconv", &isomp->im_flags, ISOFSMNT_KICONV);
 
+	if (1 == vfs_scanopt(mp->mnt_optnew, "uid", "%d", &v)) {
```
Comment (bsdjhb) [inline sys/fs/cd9660/cd9660_vfsops.c:406]: Normal FreeBSD style here would be to use `... == 1` rather than `1 == ...`

File: sys/fs/cd9660/cd9660_vnops.c [hunk 7]
```diff
@@ -140,8 +144,27 @@ cd9660_access(struct vop_access_args *ap)
 		}
 	}
 
-	return (vaccess(vp->v_type, ip->inode.iso_mode, ip->inode.iso_uid,
-	    ip->inode.iso_gid, ap->a_accmode, ap->a_cred));
+	file_mode = ip->inode.iso_mode;
+	switch (vp->v_type) {
+	case VDIR:
+		if (ip->i_mnt->im_flags & ISOFSMNT_DMASK)
+			file_mode = ip->i_mnt->im_dmask;
```
Comment (bsdjhb) [inline sys/fs/cd9660/cd9660_vnops.c:151]: Should this be `file_mode &= ip->i_mnt->im_dmask` instead of an assignment?  msdosfs uses the &

File: sys/fs/cd9660/cd9660_vnops.c [hunk 8]
```diff
@@ -140,8 +144,27 @@ cd9660_access(struct vop_access_args *ap)
...
+		break;
+	case VREG:
```
Comment (bsdjhb) [inline sys/fs/cd9660/cd9660_vnops.c:153]: FWIW, msdosfs applies the non-directory mask to all non-`VDIR` vnodes, not just `VREG`.

File: sys/fs/cd9660/cd9660_vnops.c [hunk 9]
```diff
@@ -169,9 +192,26 @@ cd9660_getattr(struct vop_getattr_args *ap)
 	vap->va_fileid	= ip->i_number;
 
 	vap->va_mode	= ip->inode.iso_mode;
+	switch (vp->v_type) {
+	case VDIR:
+		if (ip->i_mnt->im_flags & ISOFSMNT_DMASK)
```
Comment (bsdjhb) [inline sys/fs/cd9660/cd9660_vnops.c:197]: Hmm, these also assign the mask instead of anding the mask.  In the case of an image with rock-ridge attributes this can result in increased permissions which doesn't match the description in the manpage changes.

File: sbin/mount_cd9660/mount_cd9660.c (hunk 2 above)
Comment (ricardobranco777) [reply sbin/mount_cd9660/mount_cd9660.c:320]: I copied this verbatim from `mount_msdosfs.c`:

https://github.com/freebsd/freebsd-src/blob/main/sbin/mount_msdosfs/mount_msdosfs.c#L277

File: sys/fs/cd9660/cd9660_node.c (hunk 4 above)
Comment (ricardobranco777) [reply sys/fs/cd9660/cd9660_node.c:155]: I agree with the `S_IX*` part but turning off the executable bits means a sane default for me as I wouldn't expect executables from non-Rock-Ridge.

File: sys/fs/cd9660/cd9660_node.c (hunk 5 above)
Comment (ricardobranco777) [reply sys/fs/cd9660/cd9660_node.c:159]: Good catch.  I think I enabled this because when copying with the default `cp` alias (`alias cp='cp -ip'`) I ended up with non-writable files.  Will fix.

File: sys/fs/cd9660/cd9660_vnops.c (hunk 7 above)
Comment (ricardobranco777) [reply sys/fs/cd9660/cd9660_vnops.c:151]: This part is enforcing the dmask passed as option.

File: sys/fs/cd9660/cd9660_vnops.c (hunk 9 above)
Comment (ricardobranco777) [reply sys/fs/cd9660/cd9660_vnops.c:197]: Because this is about enforcing the fmask.  Will check again the whole thing in the manpage changes.

File: sys/fs/cd9660/cd9660_node.c (hunk 3 above)
Comment (ricardobranco777) [reply sys/fs/cd9660/cd9660_node.c]: Fixed.

File: sys/fs/cd9660/cd9660_node.c (hunk 4 above)
Comment (ricardobranco777) [reply sys/fs/cd9660/cd9660_node.c]: I removed the whole change of this block.

File: sys/fs/cd9660/cd9660_node.c (hunk 5 above)
Comment (ricardobranco777) [reply sys/fs/cd9660/cd9660_node.c]: Fixed.

File: sys/fs/cd9660/cd9660_vfsops.c (hunk 6 above)
Comment (ricardobranco777) [reply sys/fs/cd9660/cd9660_vfsops.c]: Fixed.

File: sys/fs/cd9660/cd9660_vnops.c (hunk 7 above)
Comment (ricardobranco777) [reply sys/fs/cd9660/cd9660_vnops.c]: Fixed.

File: sys/fs/cd9660/cd9660_vnops.c (hunk 9 above)
Comment (ricardobranco777) [reply sys/fs/cd9660/cd9660_vnops.c:195]: Fixed.

File: sys/fs/cd9660/cd9660_vnops.c (hunk 8 above)
Comment (ricardobranco777) [reply sys/fs/cd9660/cd9660_vnops.c]: Fixed.

File: sbin/mount_cd9660/mount_cd9660.c (hunk 2 above)
Comment (bsdjhb) [reply sbin/mount_cd9660/mount_cd9660.c:320]: That's fair.

File: sbin/mount_cd9660/mount_cd9660.8 [hunk 10]
```diff
@@ -73,11 +73,11 @@ version number.
 .It Fl G Ar group
 Set the group of the files in the file system to
 .Ar group .
-The default gid on non-Rockridge volumes is zero.
+The default gid on non-Rockridge volumes is zero is the group that mounted the filesystem.
```
Comment (bsdjhb) [inline sbin/mount_cd9660/mount_cd9660.8:76]: s/is zero is/ here as well

File: sys/fs/cd9660/cd9660_vnops.c [hunk 11]
```diff
@@ -140,8 +144,25 @@ cd9660_access(struct vop_access_args *ap)
 		}
 	}
 
-	return (vaccess(vp->v_type, ip->inode.iso_mode, ip->inode.iso_uid,
-	    ip->inode.iso_gid, ap->a_accmode, ap->a_cred));
+	file_mode = ip->inode.iso_mode;
+	switch (vp->v_type) {
```
Comment (bsdjhb) [inline sys/fs/cd9660/cd9660_vnops.c:148]: If you wanted, you could perhaps simplify the `switch` down to an if.

I looked at msdosfs a bit more and what it does is to set default values of its equivalent of `im_dmask` and `im_fmask` so that it can use them unconditionally without needing flags.  You could choose to do that here as well by initializing `im_dmask` and `im_fmask` to `ACCESSPERMS` in `cd9660_mount` before you check for the options.  You could them remove the `ISOFSMNT_DMASK` and `ISOFSMNT_FMASK` flags and replace this switch with something like `file_mode &= (vp->v_type == VDIR ? ip->i_mnt->im_dmask : ip->i_mnt->im_fmask;` both here and in the `VOP_GETATTR`.

What you have now is also correct, but making that change might be more concise.

**Review by bsdjhb — COMMENTED** [review]
I think the first commit looks good (perhaps modulo some simplifications I suggested).  The second change is probably ok, but probably needs a second opinion.

File: sbin/mount_cd9660/mount_cd9660.8 (hunk 10 above)
Comment (ricardobranco777) [reply sbin/mount_cd9660/mount_cd9660.8]: Fixed.

File: sys/fs/cd9660/cd9660_vnops.c (hunk 11 above)
Comment (ricardobranco777) [reply sys/fs/cd9660/cd9660_vnops.c]: Brilliant!  I followed your suggestions and successfully tested a RR & non-RR image with & without the introduced options and it seems to work.

**Comment by bsdjhb** [comment]
I've merged the first commit.  Will try to rustle up some other reviewers for the policy change in the second commit.

**Comment by ricardobranco777** [comment]
> I've merged the first commit. Will try to rustle up some other reviewers for the policy change in the second commit.

Thank you!
